### PR TITLE
    build: add a script for setting the release version in examples, charts,  and docs

### LINKED ISSUE #15749: Need a script to generate the changes for the release tag
**Is this a bug report or feature request?**
* Feature Request

**What should the feature do:**
We really need automation for generating the needed changes for the release version. Until now, we have been using a simple search-and-replace for certain `master` tags to set the release version. For example, #15596 set the tags for the 1.17 branch. However, this can lead to bugs where extra changes are made, or changes may be missed, such as #15743 where the helm chart shipped with the wrong version. 

The script can be as simple as replacing the version in a list of specific files where we expect the update. See #15596 and #15748 for the most recent locations expected to be updated. 

The script could take the expected version as an argument, such as:
```
# the first time run in the release branch, replace "master" with this beta tag
# master --> v1.18.0-beta.0
tests/scripts/set-version.sh v1.18.0-beta.0

# In later runs, the previous release would be replaced with the new patch release
# v1.18.1 --> v1.18.2
tests/scripts/set-version.sh v1.18.2
```

The person running the release would run the script, then they would open a PR with the changes. The script would not open the PR. 

**What is use case behind this feature:**
Ensure release versions are set properly in all docs and image tags.

### PR DESCRIPTION
**Pull request by obnoxxx**
**Description of changes:**  
  This is adds the first incarnation of  of a script to prepare examples and docs for a new release.

    invocation in principle: set-release-ver.sh NEW_VER
    concrete example: set-release-ver.sh v1.16.8

The originating version ios auto-detected.

**Issues resolved by this PR:** 

Resolves #15749 

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

File: tests/scripts/set-release-ver.sh [hunk 1]
```diff
@@ -0,0 +1,64 @@
+#!/usr/bin/env bash
```
Comment (BlaineEXE) [inline tests/scripts/set-release-ver.sh:1]: This is the standard set of bash script options we should use for all scripts, except in special cases. These could be important for helping the person running the script identify when an issue has occurred.

```suggestion
#!/usr/bin/env bash
set -euo pipefail
```

**Review by BlaineEXE — CHANGES_REQUESTED** [review]

File: tests/scripts/set-release-ver.sh [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+#   TO_TAG: the version tag of the new release being prepared (e. g. v1.16.8)   
+#
+# The script  replaces all occurrences of $FROM_TAG with  $TO_TAG in
+# the relevant affected files.
+
+
+# The two tags can be passed to the script as the first two positional parameters in its invocation.
+
+# So an example invocation could be:
+# set-release-ver.sh v1.16.7 v1.16.8
```
Comment (BlaineEXE) [inline tests/scripts/set-release-ver.sh:27]: Can this be made less verbose?

File: tests/scripts/set-release-ver.sh [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+
+FROM_TAG="$1"
+TO_TAG="$2"
```
Comment (BlaineEXE) [inline tests/scripts/set-release-ver.sh:31]: Personally, I prefer to put args right under the shebang and `set` line (separated by an empty line) so that it acts as its own documentation. From these lines, I can interpret immediately that there are 2 args and what they are named for the rest of the script.

**Review by BlaineEXE — CHANGES_REQUESTED** [review]

File: tests/scripts/set-release-ver.sh [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+ Documentation/Storage-Configuration/Monitoring/ceph-monitoring.md
+ Documentation/Upgrade/rook-upgrade.md
+ deploy/charts/rook-ceph/values.yaml
+ deploy/examples/direct-mount.yaml
+ deploy/examples/images.txt
+ deploy/examples/operator-openshift.yaml
+ deploy/examples/operator.yaml
+ deploy/examples/osd-purge.yaml
+ deploy/examples/toolbox-job.yaml
+ deploy/examples/toolbox-operator-image.yaml
```
Comment (BlaineEXE) [inline tests/scripts/set-release-ver.sh:48]: What's with the single space before each line here?

**Review by BlaineEXE — CHANGES_REQUESTED** [review]
I'm getting this output on macos

```
FROM_TAG: 'master'
  TO_TAG: 'v1.18.0-beta.0'
sed: 
 Documentation/Getting-Started/quickstart.md
 Documentation/Storage-Configuration/Monitoring/ceph-monitoring.md
 Documentation/Upgrade/rook-upgrade.md
 deploy/charts/rook-ceph/values.yaml
 deploy/examples/direct-mount.yaml
 deploy/examples/images.txt
 deploy/examples/operator-openshift.yaml
 deploy/examples/operator.yaml
 deploy/examples/osd-purge.yaml
 deploy/examples/toolbox-job.yaml
 deploy/examples/toolbox-operator-image.yaml
: No such file or directory
```

File: tests/scripts/set-release-ver.sh (hunk 3 above)
Comment (obnoxxx) [reply tests/scripts/set-release-ver.sh]: done

File: tests/scripts/set-release-ver.sh (hunk 4 above)
Comment (obnoxxx) [reply tests/scripts/set-release-ver.sh]: removed.

**Comment by BlaineEXE** [comment]
When I fix the script to run correctly, this is making changes that we don't want to make using `master` and `v1.18.0-beta.0` as inputs. See example below to see an undesirable change (not made in [Travis's example](https://github.com/rook/rook/pull/15596/files) along with a desirable change.

image

I am going to mark this PR as draft. Please revisit the implementation and test thoroughly to make sure it matches Travis's example. A naive sed replace `$1` with `$2` is a bit too broad. I suspect `vx.y.1`, and `vx.y.2` are going to be corner cases as well in the upgrade doc. Some examples of those are here: 
https://github.com/rook/rook/pull/15229/files
https://github.com/rook/rook/pull/15290/files

File: tests/scripts/set-release-ver.sh [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+deploy/examples/direct-mount.yaml
+deploy/examples/images.txt
+deploy/examples/operator-openshift.yaml
+deploy/examples/operator.yaml
+deploy/examples/osd-purge.yaml
+deploy/examples/toolbox-job.yaml
+deploy/examples/toolbox-operator-image.yaml
+"
+
+sed -i -e "s/$FROM_TAG/$TO_TAG/g" $FILES
```
Comment (BlaineEXE) [inline tests/scripts/set-release-ver.sh:44]: Using an array resolves the execution error I get on macos.

```suggestion
FILES=(
Documentation/Getting-Started/quickstart.md
Documentation/Storage-Configuration/Monitoring/ceph-monitoring.md
Documentation/Upgrade/rook-upgrade.md
deploy/charts/rook-ceph/values.yaml
deploy/examples/direct-mount.yaml
deploy/examples/images.txt
deploy/examples/operator-openshift.yaml
deploy/examples/operator.yaml
deploy/examples/osd-purge.yaml
deploy/examples/toolbox-job.yaml
deploy/examples/toolbox-operator-image.yaml
)

sed -i -e "s/$FROM_TAG/$TO_TAG/g" "${FILES[@]}"
```

**Comment by obnoxxx** [comment]
The `No such file or directory`  error from sed is due to the double quoting that I added around `$FILES` that I added to silence shellcheck.  But I think in this case the word splitting that the quoting prevents  is exactly what the sed invocation needs to function correctly

**Comment by obnoxxx** [comment]
@BlaineEXE  wrote:
> > The No such file or directory error from sed is due to the double quoting that I added around $FILES that I added to silence shellcheck. But I think in this case the word splitting that the quoting prevents is exactly what the sed invocation needs to function correctly
> 
> Using an array and using it as `"${FILES[@]}"` resolves the quoting issue while still ensuring any potential paths with spaces will be quoted as sed arguments.

Applied. Thanks!

File: tests/scripts/set-release-ver.sh [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+Documentation/Storage-Configuration/Monitoring/ceph-monitoring.md
+Documentation/Upgrade/rook-upgrade.md
+deploy/charts/rook-ceph/values.yaml
+deploy/examples/direct-mount.yaml
+deploy/examples/images.txt
+deploy/examples/operator-openshift.yaml
+deploy/examples/operator.yaml
+deploy/examples/osd-purge.yaml
+deploy/examples/toolbox-job.yaml
+deploy/examples/toolbox-operator-image.yaml
```
Comment (BlaineEXE) [inline tests/scripts/set-release-ver.sh:41]: Probably a future update, but let's consider `deploy/examples/*.yaml` instead of individualized yamls. It's much more frequent that we add yamls that might have images.

The benefit of the yamls is that images should always have `image:` preceding them, and they should be in some format like `rook/ceph:tag`, which can be used to prevent false positives.

File: tests/scripts/set-release-ver.sh [hunk 7]
```diff
@@ -0,0 +1,45 @@
+#!/usr/bin/env bash
+set -euo pipefail
+
+FROM_TAG="$1"
```
Comment (travisn) [inline tests/scripts/set-release-ver.sh:4]: Can we please detect the from tag automatically? It really doesn't matter what the current tag is, we just want to update whatever is there.
```suggestion
FROM_TAG=$(grep "git clone --single-branch" Documentation/Getting-Started/quickstart.md | awk '{ print $6 }')
```

File: build/release/README.md [hunk 8]
```diff
@@ -79,6 +79,18 @@ After the PR is merged, you can tag the release with the beta tag (`v1.13.0-beta
 
 **IMPORTANT** Before tagging the release, open a new PR to update the documentation and example manifest tags to the release version.
 
+The script  `set-release-ver.sh` can help preparing these changes:
+
+```console
+$ tests/scripts/set-release-ver.sh OLD_VER NEW_VER
```
Comment (travisn) [inline build/release/README.md:85]: Let's just simplify the section to show the example directly with a comment to explain the arg, then no need for lines 88-92
```suggestion
# Pass the new version
$ tests/scripts/set-release-ver.sh v1.16.8
```

File: tests/scripts/set-release-ver.sh [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+deploy/examples/direct-mount.yaml
+deploy/examples/images.txt
+deploy/examples/operator-openshift.yaml
+deploy/examples/operator.yaml
+deploy/examples/osd-purge.yaml
+deploy/examples/toolbox-job.yaml
+deploy/examples/toolbox-operator-image.yaml
+)
+
+sed -i -e "s/$FROM_TAG/$TO_TAG/g" "${FILES[@]}"
```
Comment (travisn) [inline tests/scripts/set-release-ver.sh:44]: This command will be too general. If the tag is `master`, we will end up replacing too many references. Instead, let's use very specific `sed` commands to avoid changing too many places. For example:
```
"s/docker.io/rook/ceph:$FROM_TAG/docker.io/rook/ceph:$TO_TAG/g" 
"s/tag: $FROM_TAG/tag: $TO_TAG/g" 
```
And there are others, basically we should end up with the same results as we see in a previous PR such as #15737. It should be simple to run the script and inspect that we get the same results as that PR.

**Review by travisn — CHANGES_REQUESTED** [review]

File: tests/scripts/set-release-ver.sh (hunk 7 above)
Comment (BlaineEXE) [reply tests/scripts/set-release-ver.sh:4]: I'm not sure this is so simple when we cut a `v1.Y.0-beta.0` tag from `master` when it's already been branched to `release-v1.Y`. It seems like it might be good to add autodetection later and focus first on getting the find/replace logic correct.

Comment (travisn) [reply tests/scripts/set-release-ver.sh:4]: We don't need to look at the branch name, I think it really is simple if we just extract the value in the 6th position from that grep command.

Comment (travisn) [reply tests/scripts/set-release-ver.sh:4]: To be clear, we aren't running a `git` command, we are just grepping for a line with "git clone".

File: tests/scripts/set-release-ver.sh (hunk 2 above)
Comment (obnoxxx) [reply tests/scripts/set-release-ver.sh]: @BlaineEXE 
> Can this be made less verbose?

It is a lot less verbose now.

File: tests/scripts/set-release-ver.sh (hunk 9 above)
Comment (obnoxxx) [reply tests/scripts/set-release-ver.sh:47]: @travisn  wrote:
> This command will be too general. If the tag is `master`, we will end up replacing too many references. Instead, let's use very specific `sed` commands to avoid changing too many places. For example:
> 
> ```
> "s/docker.io/rook/ceph:$FROM_TAG/docker.io/rook/ceph:$TO_TAG/g" 
> "s/tag: $FROM_TAG/tag: $TO_TAG/g" 
> ```

sounds good. I was trying to implement something along those lines.

But I think this might be too narrow: It might miss the documentation pieces.

I might end up using different sed rules for different sets of files. - wdyt?


> And there are others, basically we should end up with the same results as we see in a previous PR such as #15737. It should be simple to run the script and inspect that we get the same results as that PR.

File: tests/scripts/set-release-ver.sh [hunk 10]
```diff
@@ -0,0 +1,53 @@
+#!/usr/bin/env bash
+set -euo pipefail
+
+FROM_TAG=$(grep "git clone --single-branch" Documentation/Getting-Started/quickstart.md | awk '{ print $6 }')
```
Comment (travisn) [inline tests/scripts/set-release-ver.sh:4]: Instead of using the quickstart guide, I'm thinking it will be more reliable in the long term to parse the images.txt, something like this:
```suggestion
FROM_TAG=$(grep "docker.io/rook/ceph" deploy/examples/images.txt | awk -F : '{ print $2 }')
```

File: tests/scripts/set-release-ver.sh [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+FILES_DOCS=(
+Documentation/Getting-Started/quickstart.md
+Documentation/Storage-Configuration/Monitoring/ceph-monitoring.md
+Documentation/Upgrade/rook-upgrade.md
+)
+FILES_CHARTS=(
+deploy/charts/rook-ceph/values.yaml
+)
+FILES_EXAMPLES=(
+deploy/examples/direct-mount.yaml
```
Comment (travisn) [inline tests/scripts/set-release-ver.sh:38]: Instead of listing individual files, can we just use `deploy/examples/*`? Or does that make unexpected updates?

File: tests/scripts/set-release-ver.sh [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+FROM_TAG=$(grep "git clone --single-branch" Documentation/Getting-Started/quickstart.md | awk '{ print $6 }')
+TO_TAG="$1"
+
+# This script is supposed to facilitate the preparation of a rook release by
+# changing example and documentation files to use the new release version instead the old one.
+#
+# It will usually be run on a release branch and the resulting changes will be the content of the
+# commit for the basis of the release.
+
+# The script needs two version tags:
```
Comment (travisn) [inline tests/scripts/set-release-ver.sh:13]: Only one version tag now

File: tests/scripts/set-release-ver.sh [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+# FROM_TAG: the version tag of the previous/current release (e. g. v1.16.7)
+# TO_TAG: the version tag of the new release being prepared (e. g. v1.16.8)
+
+# FROM_TAG is auto-detected and TO_TAG should be passed to the script as the first positional parameter in its invocation.
+
+# So an example invocation could be:
+# set-release-ver.sh v1.16.8
+#
+# This assumes invocation from the release-1.16 branch with latest
+# release tag v1.16.7
```
Comment (travisn) [inline tests/scripts/set-release-ver.sh:24]: For the example, how about the next release as the example? 
```suggestion
# release tag v1.17.2
```

File: tests/scripts/set-release-ver.sh [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+#
+# It will usually be run on a release branch and the resulting changes will be the content of the
+# commit for the basis of the release.
+
+# The script needs two version tags:
+
+# FROM_TAG: the version tag of the previous/current release (e. g. v1.16.7)
+# TO_TAG: the version tag of the new release being prepared (e. g. v1.16.8)
+
+# FROM_TAG is auto-detected and TO_TAG should be passed to the script as the first positional parameter in its invocation.
```
Comment (travisn) [inline tests/scripts/set-release-ver.sh:18]: I'd suggest no need to document the FROM_TAG, it seems self explanatory. How about just add a simple comment before line 4 directly with the var?
```suggestion
# TO_TAG should be passed to the script as the first positional parameter in its invocation.
```

File: tests/scripts/set-release-ver.sh [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+deploy/examples/operator-openshift.yaml
+deploy/examples/operator.yaml
+deploy/examples/osd-purge.yaml
+deploy/examples/toolbox-job.yaml
+deploy/examples/toolbox-operator-image.yaml
+)
+
+sed -i -e "s/docker.io\/rook\/ceph:$FROM_TAG/docker.io\/rook\/ceph:$TO_TAG/g" "${FILES_EXAMPLES[@]}"
+sed -i -e  "s/tag: $FROM_TAG/tag: $TO_TAG/g" "${FILES_CHARTS[@]}"
+sed -i -e  "s/$FROM_TAG/$TO_TAG/g" "${FILES_DOCS[@]}"
```
Comment (travisn) [inline tests/scripts/set-release-ver.sh:49]: The upgrade guide will need some more special handling, to match what is done for example in #15737.

**Review by travisn — CHANGES_REQUESTED** [review]

File: tests/scripts/set-release-ver.sh (hunk 11 above)
Comment (obnoxxx) [reply tests/scripts/set-release-ver.sh:38]: will try and report back

File: tests/scripts/set-release-ver.sh (hunk 12 above)
Comment (obnoxxx) [reply tests/scripts/set-release-ver.sh]: iu ses two but only one needs to be passed as argument. Wording updated.

**Comment by obnoxxx** [comment]
@BlaineEXE  wrote:
> When I fix the script to run correctly, this is making changes that we don't want to make using `master` and `v1.18.0-beta.0` as inputs. See example below to see an undesirable change (not made in [Travis's example](https://github.com/rook/rook/pull/15596/files) along with a desirable change.
> image
> 
> I am going to mark this PR as draft. Please revisit the implementation and test thoroughly to make sure it matches Travis's example. A naive sed replace `$1` with `$2` is a bit too broad. I suspect `vx.y.1`, and `vx.y.2` are going to be corner cases as well in the upgrade doc. Some examples of those are here: https://github.com/rook/rook/pull/15229/files https://github.com/rook/rook/pull/15290/files

Thanks for pointing this out!

It took me a while to understan the problem here. 

The latest version of the script leaves those liks alone.

File: tests/scripts/set-release-ver.sh (hunk 11 above)
Comment (obnoxxx) [reply tests/scripts/set-release-ver.sh]: tested and it looks good to me. PR updated accordingly.

**Comment by obnoxxx** [comment]
I think  the broader requests and suggestions have been addressed.

The script seems to be working correctly in my local tests now as far as I can tell.

removing Draft ...

File: tests/scripts/set-release-ver.sh [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+# TO_TAG should be passed to the script as the first positional parameter in its invocation.
+
+# So an example invocation could be:
+# set-release-ver.sh v1.17.2
+#
+# This assumes invocation from the release-1.17 branch with latest
+# release tag v1.17.1
+#
+# The script replaces all occurrences of $FROM_TAG with $TO_TAG in
+# the relevant affected files.
```
Comment (BlaineEXE) [inline tests/scripts/set-release-ver.sh:27]: Can this be summarized in fewer lines? 

I can see several things that can be cleaned up:
1. FROM_TAG is no longer an input and thus doensn't need documentation
2. invocation isn't needed because it's documented in the release  readme
3. don't need to state TO_TAG is final positional because that's implicitly documented above

In addition, just some repetition that can be reduced. I think this could be 2-3 lines and still perfectly helpful

File: tests/scripts/set-release-ver.sh [hunk 17]
```diff
@@ -0,0 +1,43 @@
+#!/usr/bin/env bash
+set -euo pipefail
+
+FROM_TAG=$(grep "docker.io/rook/ceph" deploy/examples/images.txt | awk -F : '{ print $2 }')
+TO_TAG="$1"
```
Comment (BlaineEXE) [inline tests/scripts/set-release-ver.sh:5]: Putting input args at the top serves to make them more self documenting. Because FROM_TAG is no longer an input arg (but still critical to understanding the function), I would suggest making it the second section. A comment also wouldn't hurt

```suggestion
TO_TAG="$1"

FROM_TAG=$(grep "docker.io/rook/ceph" deploy/examples/images.txt | awk -F : '{ print $2 }')

```

File: tests/scripts/set-release-ver.sh [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+Documentation/Getting-Started/quickstart.md
+Documentation/Storage-Configuration/Monitoring/ceph-monitoring.md
+Documentation/Upgrade/rook-upgrade.md
+)
+FILES_CHARTS=(
+deploy/charts/rook-ceph/values.yaml
+)
+FILES_EXAMPLES=(
+    deploy/examples/*
+)
```
Comment (BlaineEXE) [inline tests/scripts/set-release-ver.sh:39]: Please pick an indentation and stick with it. Personally, indented looks best to me. I'd also suggest make sure the indentation is 2 spaces, to match other rook bash scripts like `github-action-helper.sh`.

```suggestion
FILES_DOCS=(
  Documentation/Getting-Started/quickstart.md
  Documentation/Storage-Configuration/Monitoring/ceph-monitoring.md
  Documentation/Upgrade/rook-upgrade.md
)
FILES_CHARTS=(
  deploy/charts/rook-ceph/values.yaml
)
FILES_EXAMPLES=(
  deploy/examples/*
)
```

File: tests/scripts/set-release-ver.sh [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+
+sed -i -e "s/docker.io\/rook\/ceph:$FROM_TAG/docker.io\/rook\/ceph:$TO_TAG/g" "${FILES_EXAMPLES[@]}"
+sed -i -e  "s/tag: $FROM_TAG/tag: $TO_TAG/g" "${FILES_CHARTS[@]}"
+sed -i -e  "s/ $FROM_TAG / $TO_TAG /g" "${FILES_DOCS[@]}"
```
Comment (BlaineEXE) [inline tests/scripts/set-release-ver.sh:43]: Nit: variables are declared in opposite order that sed commands are called. I'd suggest ordering them the same to help with legibility

Comment (BlaineEXE) [inline tests/scripts/set-release-ver.sh:43]: Please continue to be vigilant for unnecessary double spaces

```suggestion
sed -i -e "s/docker.io\/rook\/ceph:$FROM_TAG/docker.io\/rook\/ceph:$TO_TAG/g" "${FILES_EXAMPLES[@]}"
sed -i -e "s/tag: $FROM_TAG/tag: $TO_TAG/g" "${FILES_CHARTS[@]}"
sed -i -e "s/ $FROM_TAG / $TO_TAG /g" "${FILES_DOCS[@]}"
```

File: build/release/README.md [hunk 20]
```diff
@@ -79,6 +79,12 @@ After the PR is merged, you can tag the release with the beta tag (`v1.13.0-beta
 
 **IMPORTANT** Before tagging the release, open a new PR to update the documentation and example manifest tags to the release version.
 
+The script  `set-release-ver.sh` can help preparing these changes:
+
+```console
+# Pass the new version
+$ tests/scripts/set-release-ver.sh v1.16.8
+```
```
Comment (travisn) [inline build/release/README.md:87]: nit: add a blank line after the block

File: build/release/README.md [hunk 21]
```diff
@@ -79,6 +79,12 @@ After the PR is merged, you can tag the release with the beta tag (`v1.13.0-beta
 
 **IMPORTANT** Before tagging the release, open a new PR to update the documentation and example manifest tags to the release version.
 
+The script  `set-release-ver.sh` can help preparing these changes:
```
Comment (BlaineEXE) [inline build/release/README.md:82]: Have you looked into enabling whitespace visibility in your editor yet to help prevent double space issues? Please do this.

```suggestion
The script `set-release-ver.sh` can help preparing these changes:
```

File: tests/scripts/set-release-ver.sh (hunk 19 above)
Comment (travisn) [inline tests/scripts/set-release-ver.sh:43]: The upgrade guide will still need some more special handling, to match what is done for example in https://github.com/rook/rook/pull/15737.

File: build/release/README.md [hunk 22]
```diff
@@ -79,6 +79,12 @@ After the PR is merged, you can tag the release with the beta tag (`v1.13.0-beta
...
+
+```console
+# Pass the new version
```
Comment (BlaineEXE) [inline build/release/README.md:85]: I think "pass the new version" is better left as an instruction above. It'll be easily copy-pasted with the line below as part of the snippet. Seems best to leave it so it won't risk causing any issues when that happens

File: build/release/README.md [hunk 23]
```diff
@@ -79,6 +79,12 @@ After the PR is merged, you can tag the release with the beta tag (`v1.13.0-beta
...
+$ tests/scripts/set-release-ver.sh v1.16.8
```
Comment (BlaineEXE) [inline build/release/README.md:86]: Other console snippets above don't use dollar signs. If we leave this out copying and pasting is easier with less friction

```suggestion
tests/scripts/set-release-ver.sh v1.16.8
```

**Review by BlaineEXE — CHANGES_REQUESTED** [review]

File: tests/scripts/set-release-ver.sh [hunk 24]
```diff
@@ -0,0 +1,43 @@
+#!/usr/bin/env bash
```
Comment (travisn) [inline tests/scripts/set-release-ver.sh:1]: Have you tried running this script on your mac? It doesn't seem to work for me, the files are not updated in place by `sed`.

**Review by travisn — CHANGES_REQUESTED** [review]

File: tests/scripts/set-release-ver.sh (hunk 24 above)
Comment (BlaineEXE) [reply tests/scripts/set-release-ver.sh:1]: I was wondering when/if this would come up. Mac's sed version is POSIX/BSD sed, not gnu-sed, and the -i option is different in gnu.

This may need the `-i ''` option as noted here: https://www.baeldung.com/linux/gnu-bsd-stream-editor#2-the--i-option

File: build/release/README.md (hunk 21 above)
Comment (obnoxxx) [reply build/release/README.md]: @BlaineEXE wrote:
> Have you looked into enabling whitespace visibility in your editor yet to help prevent double space issues?

yes but no conclusion yet.


> Please do this.

I will continue and find a solution

File: build/release/README.md (hunk 20 above)
Comment (obnoxxx) [reply build/release/README.md:87]: done

File: tests/scripts/set-release-ver.sh (hunk 24 above)
Comment (obnoxxx) [reply tests/scripts/set-release-ver.sh:1]: I tried it but the main problem currently seems to be that sed refuses top operate on non-regular filel like directory  `deploy/examples/cosi` which are expanded by `deploy/examples/*`.

I will figure it uot and update.

Comment (obnoxxx) [reply tests/scripts/set-release-ver.sh:1]: FWIW: works just fine for me on Linux where I developed it initially. will take care of MacOS next.

File: tests/scripts/set-release-ver.sh (hunk 16 above)
Comment (obnoxxx) [reply tests/scripts/set-release-ver.sh]: done.

File: tests/scripts/set-release-ver.sh (hunk 19 above)
Comment (obnoxxx) [reply tests/scripts/set-release-ver.sh]: gppd point. changed.

File: tests/scripts/set-release-ver.sh (hunk 24 above)
Comment (obnoxxx) [reply tests/scripts/set-release-ver.sh:1]: MacOS works again now.

File: build/release/README.md (hunk 22 above)
Comment (obnoxxx) [reply build/release/README.md]: done

File: tests/scripts/set-release-ver.sh (hunk 11 above)
Comment (obnoxxx) [reply tests/scripts/set-release-ver.sh]: In fact I replaced it with a command `finddeploy/examples -type f`now.

File: tests/scripts/set-release-ver.sh (hunk 24 above)
Comment (obnoxxx) [reply tests/scripts/set-release-ver.sh:1]: @BlaineEXE  wrote:
> I was wondering when/if this would come up. Mac's sed version is POSIX/BSD sed, not gnu-sed, and the -i option is different in gnu.
> 
> This may need the `-i ''` option as noted here: https://www.baeldung.com/linux/gnu-bsd-stream-editor#2-the--i-option

after struggling with a correct invocation  for a while, I ended up uysing the `build/sed-in-place` script.

File: tests/scripts/set-release-ver.sh [hunk 25]
```diff
@@ -0,0 +1,28 @@
...
+
+FILES_DOCS=(
+ Documentation/Getting-Started/quickstart.md
+ Documentation/Storage-Configuration/Monitoring/ceph-monitoring.md
+ Documentation/Upgrade/rook-upgrade.md
+)
+FILES_CHARTS=(
+ deploy/charts/rook-ceph/values.yaml
+)
+mapfile -t FILES_EXAMPLES < <(find deploy/examples -type f)
```
Comment (travisn) [inline tests/scripts/set-release-ver.sh:24]:
```suggestion
```

File: tests/scripts/set-release-ver.sh [hunk 26]
```diff
@@ -0,0 +1,28 @@
...
+
+$SED "s/ $FROM_TAG / $TO_TAG /g" "${FILES_DOCS[@]}"
+$SED "s/tag: $FROM_TAG/tag: $TO_TAG/g" "${FILES_CHARTS[@]}"
+$SED "s/docker.io\/rook\/ceph:$FROM_TAG/docker.io\/rook\/ceph:$TO_TAG/g" "${FILES_EXAMPLES[@]}"
```
Comment (travisn) [inline tests/scripts/set-release-ver.sh:28]:
```suggestion
$SED "s/docker.io\/rook\/ceph:$FROM_TAG/docker.io\/rook\/ceph:$TO_TAG/g" deploy/examples/*.yaml
$SED "s/docker.io\/rook\/ceph:$FROM_TAG/docker.io\/rook\/ceph:$TO_TAG/g" deploy/examples/images.txt
```

**Review by travisn — CHANGES_REQUESTED** [review]
With these suggestions, the script is working for me, except we need to fine tune the changes to rook-upgrade.md.

File: tests/scripts/set-release-ver.sh (hunk 26 above)
Comment (BlaineEXE) [reply tests/scripts/set-release-ver.sh:32]: Alternately, I think this could be one line which could avoid duplicating the search-replace expression

```
$SED "s/docker.io\/rook\/ceph:$FROM_TAG/docker.io\/rook\/ceph:$TO_TAG/g" deploy/examples/*.yaml deploy/examples/images.txt
```

File: tests/scripts/set-release-ver.sh [hunk 27]
```diff
@@ -0,0 +1,32 @@
+#!/usr/bin/env bash
+set -euo pipefail
+
+TO_TAG="$1"
+
+FROM_TAG=$(grep "docker.io/rook/ceph" deploy/examples/images.txt | awk -F : '{ print $2 }')
+
+scriptdir="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+SED="${scriptdir}/../../build/sed-in-place"
```
Comment (BlaineEXE) [inline tests/scripts/set-release-ver.sh:9]: This raises a good point. `SED` is defined with an absolute path, which is good form. I can foresee someone running this script from the `tests/scripts` dir and then it not working correctly. (aside: @travisn should this reside next to `sed-in-place` in the build dir instead? IMO yes.)

Given that, I would suggest a different strategy that will ensure the whole script works, not just the sed portion.

```suggestion
scriptdir="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
rook_root="${scriptdir}/../..}"
cd "${rook_root}" # allow using relative paths in script

SED="build/sed-in-place"
```

File: tests/scripts/set-release-ver.sh [hunk 28]
```diff
@@ -0,0 +1,32 @@
...
+)
+FILES_CHARTS=(
+ deploy/charts/rook-ceph/values.yaml
+)
+FILES_EXAMPLES=(
+ deploy/examples/*.yaml
+ deploy/examples/images.txt
+)
+
+$SED -e "s/rook\/ceph:$FROM_TAG/rook\/ceph:$TO_TAG/g" -e "s/rook-version=$FROM_TAG/rook-version=$TO_TAG/g" Documentation/Upgrade/rook-upgrade.md
```
Comment (BlaineEXE) [inline tests/scripts/set-release-ver.sh:29]: It's easier to read separate sed expressions when they're broken up to one expression per line

```suggestion
$SED -e "s/rook\/ceph:$FROM_TAG/rook\/ceph:$TO_TAG/g" \
     -e "s/rook-version=$FROM_TAG/rook-version=$TO_TAG/g" \
  Documentation/Upgrade/rook-upgrade.md
```

**Review by BlaineEXE — CHANGES_REQUESTED** [review]
shellcheck is failing

File: tests/scripts/set-release-ver.sh [hunk 29]
```diff
@@ -0,0 +1,54 @@
...
+
+echo "Updating upgrade guide from $FROM_TAG_UPGRADE to $TO_TAG..."
+
+$SED -e "s/rook-version=$FROM_TAG_UPGRADE/rook-version=$TO_TAG/g" \
+    -e "s/rook\/ceph:$FROM_TAG_UPGRADE/rook\/ceph:$TO_TAG/g" \
+    -e "s/\`$FROM_TAG_UPGRADE\`/\`$TO_TAG\`/g" \
+    -e "s/when Rook $FROM_TAG_UPGRADE/when Rook $TO_TAG/g" \
+  Documentation/Upgrade/rook-upgrade.md
+
+echo "Done! Now you can open a PR with these changes."
```
Comment (BlaineEXE) [inline tests/scripts/set-release-ver.sh:54]: There is a corner case here when we are branching new release docs from master. In that case, we are updating both current/new version and previous/deprecated versions. Do you want to handle that case with this script? IMO, it might be good to defer this for later, but I think it might be something to at least note in a comment so we can be reminded that the script is really only fully featured for z-stream updates but not new branch upgrades.

Screenshot b/c formatting messes up when copying code with code blocks:

image

If this were the master branch and we were starting the v1.18 docs, there would be 2 updates here
1. Change v1.17.0 to either v1.18.0
2. Change v1.16.* to v1.17.Z (where Z is the latest 1.17)

**Review by BlaineEXE — APPROVED** [review]
This looks good to merge to me, with a note/question about a corner case and questions about if we want to handle that here: https://github.com/rook/rook/pull/15781/files#r2070521987

File: tests/scripts/set-release-ver.sh (hunk 29 above)
Comment (travisn) [reply tests/scripts/set-release-ver.sh:54]: I've tested the conversion when starting from the master branch, and it covers most of your concerns. The `FROM_TAG_UPGRADE` is specially designed for this case. 

> 1. Change v1.17.0 to either v1.18.0

If we assume the manual update to the upgrade guide is done before we run this script (as in #15643 for 1.17), then the next time we run this, the tag will already be v1.18.0, and the script would detect that and update it to v1.18.0-beta.0. 

> 2.  Change v1.16.* to v1.17.Z (where Z is the latest 1.17)

This is not covered by the script. If we want to update the previous minor version's patch version, it will be manual. I'm not sure that's worth automating.

Comment (BlaineEXE) [reply tests/scripts/set-release-ver.sh:54]: Sounds like the right approach to me. Thanks
